### PR TITLE
[READY] IPC's from BeeStation

### PR DESCRIPTION
**Pull request by JohnFulpWillard**
## About The Pull Request

This PR ports over Bee's IPC race and almost everything IPC related along (With many differences due to downstream limitations and other things). Since Bee's and Fulp's chems are different, I didn't want to port over the IPC-only chemicals, so instead just made them able to use all chems, and since we physically can't mess with /tg/ antag files, they can be Changelings and can be absorbed (So they can be husked). I've played on one of Bee's servers and was amazed at just how good IPC's play and RP as, and I feel disappointed that we don't have them.

Things to do:
- [ ] Finish porting all the needed sprites
- [ ] Get it working without any errors
- [ ] Make sure they can properly use chemicals, but can't eat anything
- [ ] Make sure they can get revived and check for any wacky things with their brain being in the chest

## Why It's Good For The Game

More playable races good.

## Changelog
:cl:
add: Added IPCs
add: Added IPC organs and implants in the exosuit fabricators
/:cl:

### CODE REVIEW

**Comment by ATH1909** [comment]
I will watch this PR with interest.

File: code/modules/mob/living/carbon/human/species_types/IPC.dm [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	damage_overlay_type = "synth"
+	limbs_id = "synth"
+	mutant_bodyparts = list("ipc_screen", "ipc_antenna", "ipc_chassis")
+	default_features = list("ipc_screen" = "BSOD", "ipc_antenna" = "None")
+	burnmod = 2
+	heatmod = 1.5
+	brutemod = 1
+	toxmod = 0
+	clonemod = 0
+	staminamod = 0.8
```
Comment (ATH1909) [inline code/modules/mob/living/carbon/human/species_types/IPC.dm:30]: That's a YIKES from me, dog. This, among other things, makes you take 3 stun baton strikes to fully take down.

File: code/modules/mob/living/carbon/human/species_types/IPC.dm [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+/datum/species/ipc/random_name(unique)
+	var/ipc_name = "[pick(GLOB.posibrain_names)]-[rand(100, 999)]"
+	return ipc_name
+
+/datum/species/ipc/on_species_gain(mob/living/carbon/C) // Let's make that IPC actually robotic.
+	. = ..()
+	var/obj/item/organ/appendix/appendix = C.getorganslot("appendix") // Easiest way to remove it.
+	appendix.Remove(C)
+	QDEL_NULL(appendix)
+	ADD_TRAIT(C, TRAIT_XENO_IMMUNE, "xeno immune") //makes the IPC immune to huggers
```
Comment (ATH1909) [inline code/modules/mob/living/carbon/human/species_types/IPC.dm:54]: Isn't the thing is quotes supposed to sort of describe the source of the trait, not the trait itself? I could be wrong, though.

File: code/modules/reagents/chemistry/holder.dm [hunk 3]
```diff
@@ -348,6 +348,8 @@
   * * liverless - Stops reagents that aren't set as [/datum/reagent/var/self_consuming] from metabolizing
   */
 /datum/reagents/proc/metabolize(mob/living/carbon/C, can_overdose = FALSE, liverless = FALSE)
+	if(C?.dna?.species && (NOREAGENTS in C.dna.species.species_traits))
```
Comment (ATH1909) [inline code/modules/reagents/chemistry/holder.dm:351]: mate, we already have a trait that makes you not process reagents (that aren't specifically handled by your race): TRAIT_NOMETABOLISM

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 4]
```diff
@@ -348,6 +348,22 @@
...
+/datum/reagent/medicine/system_cleaner
+	name = "System Cleaner"
+	description = "Neutralizes harmful chemical compounds inside synthetic systems."
+	reagent_state = LIQUID
+	color = "#F1C40F"
+	metabolization_rate = 0.5 * REAGENTS_METABOLISM
+	process_flags = SYNTHETIC
+
+/datum/reagent/medicine/system_cleaner/on_mob_life(mob/living/M)
+	M.adjustToxLoss(-2*REM, 0)
```
Comment (ATH1909) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm:360]: aren't both androids and IPCs immune to tox damage? what's the point of this chem?

File: code/modules/ruins/objects_and_mobs/sin_ruins.dm [hunk 5]
```diff
@@ -14,6 +14,11 @@
 		return
 	if(obj_flags & IN_USE)
 		return
+	if(isIPC(user))
+		user.visible_message("<span class='warning'> As [user] tries to pull \the [src]'s lever, the machine seems to hesitate a bit.</span>", "<span class='warning'>You feel as if you are trying to put at stake something you don't even have...\ You suddenly feel your mind... Suboptimal?</span>")
```
Comment (ATH1909) [inline code/modules/ruins/objects_and_mobs/sin_ruins.dm:18]:
```suggestion
		user.visible_message("<span class='warning'> As [user] tries to pull \the [src]'s lever, the machine seems to hesitate a bit.</span>", "<span class='warning'>You feel as if you are trying to put at stake something you don't even have...\ You suddenly feel your mind become... suboptimal?</span>")
```

File: code/modules/surgery/organs/stomach.dm [hunk 6]
```diff
@@ -162,3 +162,22 @@
...
+	desc = "A micro-cell, for IPC use only. Do not swallow."
+	status = ORGAN_ROBOTIC
+
+/obj/item/organ/stomach/cell/emp_act(severity)
+	switch(severity)
+		if(1)
+			owner.nutrition = 50
+			to_chat(owner, "<span class='warning'>Alert: Heavy EMP Detected. Rebooting power cell to prevent damage.</span>")
+		if(2)
+			owner.nutrition = 250
```
Comment (ATH1909) [inline code/modules/surgery/organs/stomach.dm:182]: me and the boys EMPing ourselves while low on nutrition to INCREASE our sub-250 nutrition values to 250:
image

File: code/modules/surgery/organs/liver.dm [hunk 7]
```diff
@@ -118,3 +118,21 @@
...
+	name = "substance processor"
+	icon_state = "substance_processor"
+	attack_verb = list("processed")
+	desc = "A machine component, installed in the chest. This grants the Machine the ability to process chemicals that enter its systems."
+	alcohol_tolerance = 0
+	toxTolerance = -1
+	toxLethality = 0
+	status = ORGAN_ROBOTIC
+
+/obj/item/organ/liver/cybernetic/upgraded/ipc/emp_act(severity)
```
Comment (ATH1909) [inline code/modules/surgery/organs/liver.dm:132]: it looks like it's in your best interesting to just ditch this thing near-roundstart as an IPC; you can't metabolize most chems anyway.

File: code/modules/surgery/organs/ears.dm [hunk 8]
```diff
@@ -119,3 +119,26 @@
...
+	status = ORGAN_ROBOTIC
+
+/obj/item/organ/ears/robot/emp_act(severity)
+	switch(severity)
+		if(1)
+			owner.Jitter(30)
+			owner.Dizzy(30)
+			owner.Knockdown(200)
+			deaf = 30
+			to_chat(owner, "<span class='warning'>Your robotic ears are ringing, uselessly.</span>")
```
Comment (ATH1909) [inline code/modules/surgery/organs/ears.dm:139]: bit of a weird comma usage

also, don't robotic ears already have an EMP effect?

Comment (ATH1909) [reply code/modules/surgery/organs/ears.dm:139]: oh, I'm thinking of cybernetic ears, not robotic ones.

File: code/modules/mob/living/carbon/human/species_types/IPC.dm [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+			break
+		if(H.nutrition > NUTRITION_LEVEL_WELL_FED)
+			to_chat(H, "<span class='notice'>You are now fully charged.</span>")
+			break
+	H.visible_message("<span class='notice'>[H] unplugs from the [A].</span>", "<span class='notice'>You unplug from the [A].</span>")
+
+/datum/species/ipc/spec_life(mob/living/carbon/human/H)
+	. = ..()
+	if(H.health <= HEALTH_THRESHOLD_CRIT && H.stat != DEAD) // So they die eventually instead of being stuck in crit limbo.
+		H.adjustFireLoss(6) // After bodypart_robotic resistance this is ~2/second
```
Comment (ATH1909) [inline code/modules/mob/living/carbon/human/species_types/IPC.dm:159]: uh, are you sure that these guys actually get the augmented limb damage reduction stuff? also, will this get multiplied by the 2x burnmod that IPCs get, or does burnmod not get applied to direct calls to adjustFireLoss?

**Comment by ATH1909** [comment]
aw

**Comment by JohnFulpWillard** [comment]
> 
> 
> aw

Yeah, I don't think I can pull this off. There's too many unique things IPC's have that require overwriting /tg/ code, and they still use old-version type chems for IPCs. It would be better to change how they act in many ways and PR it to the upstream instead

**Pull request closed without merging**